src/api/app/models/bs_request_action.rb

```diff
@@ -919,33 +919,24 @@ def expand_targets(ignore_build_state)
     return
   end
 
-  def source_access_check!
-    sp = Package.find_by_project_and_name(source_project, source_package)
-    if sp.nil?
-      # either not there or read permission problem
-      if Package.exists_on_backend?(source_package, source_project)
-        # user is not allowed to read the source, but when he can write
-        # the target, the request creator (who must have permissions to read source)
-        # wanted the target owner to review it
-        tprj = Project.find_by_name(target_project)
-        if tprj.nil? || !User.current.can_modify_project?(tprj)
-          # produce an error for the source
-          Package.get_by_project_and_name(source_project, source_package)
-        end
-        return
-      end
-      if Project.exists_by_name(source_project)
-        # it is a remote project
-        return
+  def is_target_maintainer?(user = User.current)
+    if target_package
+      tpkg = Package.find_by_project_and_name(target_project, target_package)
+      if tpkg
+        return user.can_modify_package?(tpkg)
       end
-      # produce the same exception for webui
-      Package.get_by_project_and_name(source_project, source_package)
     end
-    if sp.class == String
-      # a remote package
-      return
-    end
-    sp.check_source_access!
+    tprj = Project.find_by_name(target_project)
+    tprj && user.can_modify_project?(tprj)
+  end
+
+  def source_access_check!
+    # check if we are maintainer in the target, so we are allowed
+    # to see the diff, since the request could not have been constructed
+    return if is_target_maintainer?
+
+    # not being a maintainer in target, run standard permission check
+    Package.get_by_project_and_name(source_project, source_package, {use_source: true})
```

Comment:
IIRC it was coded that it was only checking for remote case, no matter of target maintainer and may even skipping checks independ of remote when user is not set.

However, sourcediff is a generic thing and I wonder why webui should show a different diff.

Reply:
> IIRC it was coded that it was only checking for remote case, no matter of target maintainer 

Yes because [creating the request](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/bs_request_permission_check.rb#L160) is already checking access to the target project in various ways (in [check_action_target](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/bs_request_permission_check.rb#L139) and also in [set_permissions_for_action](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/models/bs_request_permission_check.rb#L171)). Creating requests additionally checks access to the source package with the method we are talking about here.

However #3597 is **ONLY** about [generating the diff](https://github.com/openSUSE/open-build-service/blob/master/src/api/app/mixins/request_source_diff.rb#L16). That method **ONLY** checks access to the source package and **ONLY** in this case we skip the access check **ONLY** if it's a package in a remote `Project`.